src/services/sync/strategies/LDAPSyncerConsumer.js

```diff
@@ -0,0 +1,160 @@
+			const foundUser = await UserRepo.findByLdapIdAndSchool(userData.ldapId, school._id);
+			try {
+				if (foundUser !== null) {
+					await this.updateUserAndAccount(foundUser, userData, accountData);
+				} else {
+					await this.createUserAndAccount(userData, accountData, school._id);
+				}
+			} catch (err) {
+				logger.error('LDAP SYNC: error while update or add a user', { err, syncId });
+				throw err;
+			}
+			return true;
+		}
+		return false;
+	}
+
+	async updateUserAndAccount(user, userData, account) {
+		const updateObject = {};
+		if (userData.firstName !== user.firstName) {
+			updateObject.firstName = userData.firstName || ' ';
+		}
+		if (userData.lastName !== user.lastName) {
+			updateObject.lastName = userData.lastName;
+		}
+		// Updating SchoolId will cause an issue. We need to discuss about it
+		if (userData.email !== user.email) {
+			updateObject.email = userData.email;
+		}
+		if (userData.ldapDn !== user.ldapDn) {
+			updateObject.ldapDn = userData.ldapDn;
+		}
+		// Role
+		const userRoles = user.roles && user.roles.map((r) => r.name);
+		if (!_.isEqual(userRoles, userData.roles)) {
+			updateObject.roles = userData.roles;
+		}
+		if (!_.isEmpty(updateObject)) {
+			return UserRepo.updateUserAndAccount(user._id, updateObject, account);
+		}
+		return true;
+	}
+
+	async createUserAndAccount(idmUser, account, schoolId) {
+		try {
+			idmUser.schoolId = schoolId;
+			return UserRepo.createUserAndAccount(idmUser, account);
+		} catch (err) {
+			logger.error('LDAP SYNC: error while creating User', err);
+			throw err;
+		}
+	}
+
+	async classAction(classData) {
+		const school = await SchoolRepo.findSchoolByLdapIdAndSystem(classData.schoolDn, classData.systemId);
+
+		if (school !== null) {
+			const existingClass = await ClassRepo.findClassByYearAndLdapDn(school.currentYear, classData.ldapDN);
+			try {
+				if (existingClass !== null) {
+					if (existingClass.name !== classData.name) {
+						await ClassRepo.updateClassName(existingClass._id, classData.name);
+					}
+				} else {
+					const newClass = {
+						name: classData.name,
+						schoolId: school._id,
+						nameFormat: 'static',
+						ldapDN: classData.ldapDN,
+						year: school.currentYear,
+					};
+					await ClassRepo.createClass(newClass);
+				}
+			} catch (err) {
+				logger.error('LDAP SYNC: error while update or add a class', { err, syncId: classData.syncId });
+				throw err;
+			}
+
+			return true;
+		}
+		return false;
+	}
+}
+
+const setupConsumer = () => {
+	const syncQueue = getChannel(LDAP_SYNC_CHANNEL_NAME, { durable: true });
+	const consumer = new LDAPSyncerConsumer();
+
+	const handleMessage = (incomingMessage) =>
+		consumer
+			.executeMessage(incomingMessage)
+			.then(() => {
+				return true;
+			})
+			.catch((err) => {
+				logger.error('LDAP SYNC: error while handling Stuff', { err, syncId: incomingMessage.syncId });
+				return false;
+			})
+			.finally(() => syncQueue.ackMessage(incomingMessage));
```

Comment:
I am not sure, if we should ack all messenges here. Depending on the error we could also reject the message and for example requeue them?

Reply:
We plan an fire and forget strategy. Especially if a task go wrong. The task it's irrelevant. The reason why the task is irrelevant is the sync self runs at short time periods and reproduce the task again.